src/lib.rs

```diff
@@ -49,557 +124,1353 @@ pub fn parse_document(buffer: &str) -> Node {
+            NodeType::CMarkNodeHtmlBlock => Ok(Node::HtmlBlock(HtmlBlock {resource})),
+            NodeType::CMarkNodeCustomBlock => Ok(Node::CustomBlock(CustomBlock {resource})),
+            NodeType::CMarkNodeHeading => Ok(Node::Heading(Heading {resource})),
+            NodeType::CMarkNodeThematicBreak => Ok(Node::ThematicBreak(ThematicBreak {resource})),
+            NodeType::CMarkNodeSoftbreak => Ok(Node::SoftBreak(SoftBreak {resource})),
+            NodeType::CMarkNodeLinebreak => Ok(Node::LineBreak(LineBreak {resource})),
+            NodeType::CMarkNodeCode => Ok(Node::Code(Code {resource})),
+            NodeType::CMarkNodeHtmlInline => Ok(Node::HtmlInline(HtmlInline {resource})),
+            NodeType::CMarkNodeCustomInline => Ok(Node::CustomInline(CustomeInline {resource})),
+            NodeType::CMarkNodeEmph => Ok(Node::Emph(Emph {resource})),
+            NodeType::CMarkNodeStrong => Ok(Node::Strong(Strong {resource})),
+            NodeType::CMarkNodeLink => Ok(Node::Link(Link {resource})),
+            NodeType::CMarkNodeImage => Ok(Node::Image(Image {resource})),
+        }
+    }
+
+    /// Returns the equivalent of a libcmark NodeType enum
+    pub fn get_cmark_type(&self) -> DoogieResult<NodeType> {
+        unsafe {
+            let t = cmark_node_get_type(self.as_resource().pointer);
+            Ok(NodeType::try_from(t as u32)?)
+        }
+    }
+
+    /// Returns the underlying resource object for this Node
+    fn as_resource(&self) -> Resource {
+        match self {
+            &Node::Document(ref node) => node.resource.clone(),
+            &Node::BlockQuote(ref node) => node.resource.clone(),
+            &Node::Text(ref node) => node.resource.clone(),
+            &Node::List(ref node) => node.resource.clone(),
+            &Node::Item(ref node) => node.resource.clone(),
+            &Node::CodeBlock(ref node) => node.resource.clone(),
+            &Node::HtmlBlock(ref node) => node.resource.clone(),
+            &Node::CustomBlock(ref node) => node.resource.clone(),
+            &Node::Paragraph(ref node) => node.resource.clone(),
+            &Node::Heading(ref node) => node.resource.clone(),
+            &Node::ThematicBreak(ref node) => node.resource.clone(),
+            &Node::SoftBreak(ref node) => node.resource.clone(),
+            &Node::LineBreak(ref node) => node.resource.clone(),
+            &Node::Code(ref node) => node.resource.clone(),
+            &Node::HtmlInline(ref node) => node.resource.clone(),
+            &Node::CustomInline(ref node) => node.resource.clone(),
+            &Node::Emph(ref node) => node.resource.clone(),
+            &Node::Strong(ref node) => node.resource.clone(),
+            &Node::Link(ref node) => node.resource.clone(),
+            &Node::Image(ref node) => node.resource.clone(),
+        }
+    }
+
+    /// Returns a string version of the Node type
+    pub fn get_cmark_type_string(&self) -> DoogieResult<String> {
+        unsafe {
```

Comment:
Let this comment serve as a reminder to see if you can narrow the unsafe blocks in the rest of the file.